### ISSUE TITLE
Investigate Mapbox API usage to inform billing

### ISSUE THREAD

**Issue by jilladams** [department-of-veterans-affairs/va.gov-cms#19984]
## Description 

Via #19201 we learned that our Mapbox Temporary Geocoding API usage aligns with Facility Locator search calls.

Our monthly invoicing includes calls to many other APIs, and we'd like to try to align those to usage to determine whether:
1. Calls being made are useful / necessary
2. Calls are being made from Sitewide products

### Usage
API | September 2024 usage | October 2024 usage | Nov 2024 | Dec 2024
--- | --- | --- | --- | ---
Permanent Geocoding API | 4 requests | 9 requests | 18 requests  |  7 requests
Temporary Geocoding API | 1,421,911 requests | 1,416,846 requests |  1,309,055 requests | 1,262,202 requests
Map Loads for Web | 837,440 loads | 871,572 loads | 789,351 loads | 689,438 loads
Static Images API | 1,053,302 requests | 1,168,345 requests | 1,046,238 requests | 976,666 requests
Matrix API | | 130,356 matrix elements |  156,569 matrix elements | 139,041 matrix elements

## task list, maybe
- Check content-build and vets-website code for Mapbox calls
- Gather a list of which kinds of pages make which kinds of calls
- Ask Product to help compare those pages to analytics / traffic, in order to know if our usage amounts to the API call volume
- **Product:** next step will be to determine if there are more cost-efficient ways to deliver the same functionality. (e.g. build nearby lists in Drupal, vs. via mapbox, etc)

## ACs

- [x] Document which pages make which API calls in the [Mapbox docs](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/facilities/facility-locator/engineering/data-and-api-integrations/mapbox) (which are in FL, and may need to move later)
- [x] Schedule time to review findings with PM/PO

**Comment by jilladams** [comment]
From product sync: 
* We don't think this ticket blocks autocomplete next steps.
* We need engineering to help assess where we make calls.
* Product will then need to map those uses to analytics to know the multiplier for which calls end up being expensive in terms of page views or number of times we're making the calls.
* Then: we can figure out where we can be more efficient -- alternate ways to achieve the same functionality with fewer Mapbox calls, etc.

**Comment by Agile6MSkinner** [comment]
Per @eselkin we don't need to expect multiple Mapbox tokens. There is one universal token used.

**Comment by randimays** [comment]
@Agile6MSkinner I've completed the research for Mapbox usage. [Here's the Google doc](https://docs.google.com/document/d/1rZuTa0O9oagDaWfUZGpftQHirGWQDUnIrDesAZCIexI/edit?tab=t.0 ) for my full notes for the non-Sitewide products and [the updated Product documentation](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/facilities/facility-locator/engineering/data-and-api-integrations/mapbox/mapbox-implementation.md). I didn't find any usage of the Matrix API at all in any of the Sitewide or non-Sitewide products; I'm not sure where this data is coming from. Would you like to schedule a call to go over the findings, or review async and ask questions where needed?

**Comment by randimays** [comment]
Update: I found where the Matrix API was used--it's only in Facility Locator, not in non-Sitewide products. [Product documentation](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/facilities/facility-locator/engineering/data-and-api-integrations/mapbox/mapbox-implementation.md) has been updated.

**Comment by jilladams** [comment]
@randimays Thank you!! I'm attaching a PDF here of the Google doc, as a point in time version:  [Mapbox API Usage by non-Sitewide products.pdf](https://github.com/user-attachments/files/18324710/Mapbox.API.Usage.by.non-Sitewide.products.pdf)

The Google doc steps through non-sitewide usage. 
I think we are _also_ interested in opportunities to change our own usage on Sitewide to reduce cost. For example, we are pinging Mapbox to pull static images on every page load where we hit the Static Images API. In theory we could cache those images, and stop pulling every time, and reduce our cost by quite a bit. 

While you were looking through usage / calls, not sure if you looked through the Facilities usage for opportunities like this? Is that reasonable to take a pass at?

**Comment by jilladams** [comment]
Other notes for @Agile6MSkinner @mmiddaugh 

## Cost implications re: address autosuggest
We used monthly Geocoding API usage as the baseline to estimate the 5-7x cost for using address autosuggest. 

Since Randi's findings show that not all Geocoding traffic is ours, that means the 5-7x of our own usage, and therefore the address autosuggest cost, will be lower than we estimated. (Not sure by how much, since we don't know what proportion of Geocoding API usage comes from the other 4 apps she found.)

## Wrangling other teams' usage
@mmiddaugh we should figure out how we want to tackle this. Currently, there's no way I know of to assess how much usage is coming from each other app listed, without some additional work. The options seem like (in some kind of pragmatic / doability order): 

### 1. Ask teams to move to their own Mapbox accounts.
The 4 teams in question would need to: 

1. Get COR approval for their expense
2. Create Mapbox accounts and set up their own API access tokens
3. Move their apps to use the new API access

It might be slightly more complicated than that, with 5 different accounts to hit Mapbox from the same domain. I'm not totally sure, but we could help figure it out if this is preferred. 

### 2. Try to move the account to VA
A la Mural / AWS / Figma, can VA take on the Mapbox account? Would that need to route through DOTS? 

### 3. Sitewide continues to eat the cost
We just leave it til it's a bigger issue. ?

### 4. Keep the billing on Sitewide and bill through to other teams
I think it's possible for us to set up other API access tokens within our existing Mapbox account. 
We would still need the other teams to use the new tokens, we would get their billing, and would need to figure out who / how to invoice VA for the materials expense being consumed by those other contracts. 

This sounds like a billing nightmare, and by the time those teams update their tokens, they might as well get their own accounts, I think.

**Comment by randimays** [comment]
@jilladams Ahh ok, I thought the primary purpose of this ticket was to identify the non-Sitewide applications that use Mapbox. Gotcha! I'll look into FL usage and possible optimizations as well.

**Comment by randimays** [comment]
@jilladams @Agile6MSkinner I've added some more information to the [Google doc](https://docs.google.com/document/d/1rZuTa0O9oagDaWfUZGpftQHirGWQDUnIrDesAZCIexI/edit?tab=t.0).

Point-in-time PDF: [Mapbox API Usage by non-Sitewide products - Google Docs.pdf](https://github.com/user-attachments/files/18352212/Mapbox.API.Usage.by.non-Sitewide.products.-.Google.Docs.pdf)

We have some optimization opportunities in the Static Images API and the Matrix API (which are both billable per-request, as I understand). I'm not sure exactly what the Permanent Geocoding billables are, especially seeing as how we only have a handful of requests for September and October (per the ticket description). We use the geocoding v5 endpoint extensively ... is that separately billable / do we have stats for that?

**Comment by jilladams** [comment]
Sweet! I scheduled a meeting for next Monday to step through and make sure Product knows everything they need to know in order to determine next steps. I'd like to leave this ticket open til then just to catch follow up, but we can close it after that meeting.

**Comment by randimays** [comment]
@jilladams Noting that Eli called out that we're using the Temporary Geocoding API. Not sure where the Permanent Geocoding API calls are being made but it may be helpful to drop some recent data for Temp Geocoding here.

**Comment by jilladams** [comment]
Yes, good callout, thanks. Added in Temp Geocoding API (2nd line in table, in ticket body), and added on Nov / Dec data for everything, just to see a bit more of an arc about usage.

**Comment by randimays** [comment]
We have an existing ticket for using Drupal as the source of truth for the map thumbnails on VAMC/Vet Center pages: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/15672

**Comment by randimays** [comment]
I made a few edits to the Google doc for research: 

- [Mapbox API Usage by non-Sitewide products - Google Docs.pdf](https://github.com/user-attachments/files/18414945/Mapbox.API.Usage.by.non-Sitewide.products.-.Google.Docs.pdf)
- https://docs.google.com/document/d/1rZuTa0O9oagDaWfUZGpftQHirGWQDUnIrDesAZCIexI/edit?tab=t.0

Notably: we can add additional caching time in our code to store responses from Mapbox longer than their default times. For instance, Temporary Geocoding API data is stored for five days. If you search for your zip code on Monday for VA health facilities, then again on the following Friday, you'll retrieve cached data from your search on Monday. We can store the data retrieved on that Monday for however long we want (for example, 2 weeks) to extend the time before making a fresh Mapbox call.

Eli asked about Mapbox-server-side caching (shared request caching), which would mean that multiple users seeking data for the same search could pull from cache. I don't think this is available based on the Mapbox documentation, but I think it would be worth asking our contacts at Mapbox if it is possible since it would amount to significant savings on our end.

@jilladams @Agile6MSkinner @mmiddaugh Are there any other technical things I can look into for the purposes of this spike, or do we have the information we need to move forward?

**Comment by randimays** [comment]
End-of-sprint update: I believe we are finished with the technical research for this ticket; @jilladams @Agile6MSkinner @mmiddaugh if yall agree, please close the ticket when ready.

**Comment by randimays** [comment]
Checked with @Agile6MSkinner in Slack DMs - we're good to close here.
